tozny/resource_realm_application.go

```diff
@@ -0,0 +1,170 @@
+package tozny
+
+import (
+  "context"
+  "strings"
+
+  "github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+  "github.com/tozny/e3db-clients-go/identityClient"
+)
+
+// resourceRealmApplication returns the schema and methods for provisioning a Tozny Realm Application
+func resourceRealmApplication() *schema.Resource {
+  return &schema.Resource{
+    CreateContext: resourceRealmApplicationCreate,
+    ReadContext:   resourceRealmApplicationRead,
+    DeleteContext: resourceRealmApplicationDelete,
+    Schema: map[string]*schema.Schema{
+      "client_credentials_filepath": {
+        Description: "The filepath to Tozny client credentials for the provider to use when provisioning this application.",
+        Type:        schema.TypeString,
+        Optional:    true,
+        ForceNew:    true,
+      },
+      "realm_name": {
+        Description: "The name of the Realm to provision the Application for.",
+        Type:        schema.TypeString,
+        Required:    true,
+        ForceNew:    true,
+      },
+      "client_id": {
+        Description: "The external id for clients to reference when communicating with this application.",
```

Comment:
Is there a way to validate that this is in the allowed format. I know by pratice we use lowercase-seperated-with-hyphens, but is that a keycloak requirements and should it get enforced by the provider.

Reply:
There is a way to validate the format of inputs, but for client id it can literally be anything that is a string as far as Keycloak is concerned, so not knowing all the use cases for why one would not want to follow our sensible Tozny internal I'll hold off on enforcing that standard upon our external users.